independent-projects/arc/runtime/src/main/java/io/quarkus/arc/LifecycleCallbackInvocationContext.java

```diff
@@ -0,0 +1,64 @@
+package io.quarkus.arc;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A simple InvocationContext implementation used for PostConstruct and PreDestroy callbacks.
+ */
+class LifecycleCallbackInvocationContext extends AbstractInvocationContext {
+
+    private volatile int position;
```

Comment:
why the need for `volatile` here?

I'd assume this class is not meant to be threadsafe; if it's meant to, perhaps it should use AtomicInteger instead?

Reply:
So no impl of IC in ArC is meant to be threadsafe but it should be possible to continue interception on a different thread, ie. we should safely propagate the values. 

WRT `AtomicInteger` - I'm personally a fan of atomic constructs but there are other team members who say that the fewer allocations the better ;-).

Reply:
> it should be possible to continue interception on a different thread

I thought we wanted that only for around invoke so as to minimize new object creation during interceptions?

Reply:
That's right. For now we only need this for AroundInvoke interceptors. I'll remove the `volatile` modifier and make it clear in javadoc that all lifecycle callback interceptors must be invoked on the same thread.

Reply:
> I'm personally a fan of atomic constructs but there are other team members who say that the fewer allocations the better ;-).

The fewer allocations the better! But the increment in this case is not atomic, so if it's meant to work across threads the current pattern isn't safe enough.

Reply:
As I stated elsewhere, you can definitely safely continue invocation in another thread without `volatile`.

Reply:
> But the increment in this case is not atomic, so if it's meant to work across threads the current pattern isn't safe enough.

We don't need atomicity. The IC must not be used concurrently.